split/split.1
.Dd December 2008
.Dt SPLIT 1

.Sh NAME

.Nm split
.Nd split files into pieces

.Sh SYNOPSIS

.Nm split
.Ar [-l line_count][-a suffix_length][file[name]]

.Nm split
.Ar -b n[k|m][-a suffix_length][file[name]]

.Sh DESCRIPTION

    The split utility shall read an input file and write one or more output
files. The default size of each output file shall be 1000 lines. The size of
the output files can be modified by specification of the -b or -l options.
Each output file shall be created with a unique suffix. The suffix shall
consist of exactly suffix_length lowercase letters from the POSIX locale. The
letters of the suffix shall be used as if they were a base-26 digit system,
with the first suffix to be created consisting of all 'a' characters, the
second with a 'b' replacing the last 'a', and so on, until a name of all 'z'
characters is created. By default, the names of the output files shall be
'x', followed by a two-character suffix from the character set as described
above, starting with "aa", "ab", "ac", and so on, and continuing until the
suffix "zz", for a maximum of 676 files.

    If the number of files required exceeds the maximum allowed by the suffix
length provided, such that the last allowable file would be larger than the
requested size, the split utility shall fail after creating the last file
with a valid suffix; split shall not delete the files it created with valid
suffixes. If the file limit is not exceeded, the last file created shall
contain the remainder of the input file, and may be smaller than the
requested size.

.Sh OPTIONS

    The split utility shall conform to the Base Definitions volume of IEEE
Std 1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -a  suffix_length
        Use suffix_length letters to form the suffix portion of the filenames
of the split file. If -a is not specified, the default suffix length shall be
two. If the sum of the name operand and the suffix_length option-argument
would create a filename exceeding {NAME_MAX} bytes, an error shall result;
split shall exit with a diagnostic message and no files shall be created.
    -b  n
        Split a file into pieces n bytes in size.
    -b  nk
        Split a file into pieces n*1024 bytes in size.
    -b  nm
        Split a file into pieces n*1048576 bytes in size.
    -l  line_count
        Specify the number of lines in each resulting file piece. The
line_count argument is an unsigned decimal integer. The default is 1000. If
the input does not end with a <newline>, the partial line shall be included
in the last output file.

.Sh OPERANDS

    The following operands shall be supported:

    file
        The pathname of the ordinary file to be split. If no input file is
given or file is '-', the standard input shall be used.
    name
        The prefix to be used for each of the files resulting from the split
operation. If no name argument is given, 'x' shall be used as the prefix of
the output files. The combined length of the basename of prefix and
suffix_length cannot exceed {NAME_MAX} bytes. See the OPTIONS section.

.Sh STDIN

    See the INPUT FILES section.

.Sh INPUT FILES

    Any file can be used as input.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of split:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments and input files).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    Not used.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    The output files contain portions of the original input file; otherwise,
unchanged.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        Successful completion.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    None.

.Sh EXAMPLES

    In the following examples foo is a text file that contains 5000 lines.

        Create five files, xaa, xab, xac, xad, and xae:

        split foo

        Create five files, but the suffixed portion of the created files
consists of three letters, xaaa, xaab, xaac, xaad, and xaae:

        split -a 3 foo

        Create three files with four-letter suffixes and a supplied prefix,
bar_aaaa, bar_aaab, and bar_aaac:

        split -a 4 -l 2000 foo bar_

        Create as many files as are necessary to contain at most 20*1024
bytes, each with the default prefix of x and a five-letter suffix:

        split -a 5 -b 20k foo

.Sh RATIONALE

    The -b option was added to provide a mechanism for splitting files other
than by lines. While most uses of the -b option are for transmitting files
over networks, some believed it would have additional uses.

    The -a option was added to overcome the limitation of being able to
create only 676 files.

    Consideration was given to deleting this utility, using the rationale
that the functionality provided by this utility is available via the csplit
utility (see csplit). Upon reconsideration of the purpose of the User
Portability Extension, it was decided to retain both this utility and the
csplit utility because users use both utilities and have historical
expectations of their behavior. Furthermore, the splitting on byte boundaries
in split cannot be duplicated with the historical csplit.

    The text " split shall not delete the files it created with valid
suffixes" would normally be assumed, but since the related utility, csplit,
does delete files under some circumstances, the historical behavior of split
is made explicit to avoid misinterpretation.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    csplit



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



split.c
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#ifndef _GNU_SOURCE
#define _GNU_SOURCE		/* for O_DIRECTORY, fgets_unlocked(3) */
#endif
#define __STDC_FORMAT_MACROS 1

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <fcntl.h>
#include <ctype.h>
#include <limits.h>
#include <libpu.h>
#include <assert.h>


static const char doc[] =
N_("split - split a file into pieces");

static const char args_doc[] = N_("[file [prefix]]");

static struct argp_option options[] = {
	{ "suffix-length", 'a', "N", 0,
	  N_("Use N letters to form the suffix portion of the filenames of the split file") },
	{ "bytes", 'b', "N[km]", 0,
	  N_("Split a file into pieces n bytes in size") },
	{ "lines", 'l', "N", 0,
	  N_("Specify the number of lines in each resulting file piece") },
	{ }
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

enum piece_mode_t {
	MODE_NONE,
	MODE_BYTE,
	MODE_LINE,
};


static char			inbuf[8192];
static const char		*err_fn;
static char			*suffix;

static uintmax_t		output_count;
static int			output_fd = -1;
static char			*output_fn;

static int			opt_suffix_len	= 2;
static uintmax_t		opt_piece_size	= 1000;
static enum piece_mode_t	opt_piece_mode	= MODE_LINE;
static const char		*opt_prefix	= "x";
static const char		*opt_input_fn;


static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
	case 'a': {
		int rc, tmp;

		rc = sscanf(arg, "%d", &tmp);
		if ((rc != 1) || (tmp < 1) || (tmp >= NAME_MAX))
			return ARGP_ERR_UNKNOWN;

		opt_suffix_len = tmp;
		break;
	}

	case 'b': {
		int rc;
		char ch;
		uintmax_t tmp, mult;

		rc = sscanf(arg, "%" PRIuMAX "%c", &tmp, &ch);
		if ((rc < 1) || (rc > 2) || (tmp < 1))
			return ARGP_ERR_UNKNOWN;
		if ((rc > 1) && (ch != 'k') && (ch != 'm'))
			return ARGP_ERR_UNKNOWN;
		if (rc == 1)
			ch = 0;
		else
			ch = tolower(ch);

		switch (tolower(ch)) {
		case 'k': mult = 1024; break;
		case 'm': mult = 1024 * 1024; break;
		default: mult = 1; break;
		}

		opt_piece_mode = MODE_BYTE;
		opt_piece_size = tmp * mult;
		break;
	}

	case 'l': {
		int rc;
		uintmax_t tmp;

		rc = sscanf(arg, "%" PRIuMAX, &tmp);
		if ((rc != 1) || (tmp < 1))
			return ARGP_ERR_UNKNOWN;

		opt_piece_mode = MODE_LINE;
		opt_piece_size = tmp;
		break;
	}

	case ARGP_KEY_ARG:
		switch(state->arg_num) {
		case 0:		opt_input_fn = arg; break;
		case 1:		opt_prefix = arg; break;
		default:	argp_usage (state); break; /* too many args */
		}
		break;

	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

static int incr_suffix(void)
{
	int i;

	if (!suffix) {
		suffix = (char *) xmalloc(opt_suffix_len + 1);
		for (i = 0; i < opt_suffix_len; i++)
			suffix[i] = 'a';
		suffix[opt_suffix_len] = 0;
		return 0;
	}

	for (i = opt_suffix_len - 1; i >= 0; i--) {
		if (suffix[i] != 'z') {
			suffix[i]++;
			return 0;
		}

		suffix[i] = 'a';
	}

	return 1;
}

static int open_output(void)
{
	int rc;

	if (output_fd < 0) {
		rc = incr_suffix();
		if (rc)
			return rc;

		assert(output_fn == NULL);
		output_fn = (char *) xmalloc(opt_suffix_len + strlen(opt_prefix) + 1);
		sprintf(output_fn, "%s%s", opt_prefix, suffix);

		output_fd = open(output_fn, O_WRONLY | O_CREAT | O_TRUNC, 0666);
		if (output_fd < 0) {
			perror(output_fn);
			return 1;
		}

		/* Do we care about the return value of posix_fadvise(2) ? */
		posix_fadvise(output_fd, 0, 0, POSIX_FADV_SEQUENTIAL);
	}

	return 0;
}

static int close_output(void)
{
	int rc = 0;

	if (output_fd < 0)
		return 0;

	if (close(output_fd) < 0) {
		perror(output_fn);
		rc = 1;
	}
	free(output_fn);

	output_fd = -1;
	output_fn = NULL;
	output_count = 0;

	return rc;
}

static int incr_output(uintmax_t incr)
{
	int rc = 0;

	output_count += incr;
	assert(output_count <= opt_piece_size);

	if (output_count == opt_piece_size)
		rc = close_output();

	return rc;
}

static int output_bytes(const char *buf, size_t len)
{
	int rc;
	uintmax_t dist;
	size_t wlen;

	while (len > 0) {
		rc = open_output();
		if (rc)
			return rc;

		dist = opt_piece_size - output_count;
		if (dist < len)
			wlen = dist;
		else
			wlen = len;

		rc = write_fd(output_fd, buf, wlen, output_fn);
		if (rc)
			return rc;

		buf += wlen;
		assert(wlen <= len);
		len -= wlen;

		rc = incr_output(wlen);
		if (rc)
			return rc;
	}

	return 0;
}

static int output_line(const char *s)
{
	int rc = open_output();
	if (rc)
		return rc;

	rc = write_fd(output_fd, s, strlen(s), output_fn);
	if (rc)
		return rc;

	return incr_output(1);
}

static int split_bytes(FILE *f)
{
	int done = 0;

	if (setvbuf(f, NULL, _IONBF, 0)) {
		perror(_("setvbuf(3) failed"));
		return 1;
	}

	while (!done) {
		size_t n;
		int rc;

		n = fread(inbuf, 1, sizeof(inbuf), f);
		if (ferror(f))
			return 1;
		if (feof(f))
			done = 1;

		rc = output_bytes(inbuf, n);
		if (rc)
			return rc;
	}

	return 0;
}

static int split_lines(FILE *f)
{
	char *s;

	setlinebuf(f);

	while ((s = fgets_unlocked(inbuf, sizeof(inbuf), f)) != NULL) {
		int rc = output_line(s);
		if (rc)
			return rc;
	}

	return 0;
}

static int execute_split(void)
{
	FILE *f;
	int rc = 0;

	if ((opt_input_fn == NULL) || (!strcmp(opt_input_fn, "-"))) {
		f = stdin;
		err_fn = _("(standard input)");
	} else {
		err_fn = opt_input_fn;
		if (ro_file_open(&f, opt_input_fn))
			return 1;
	}

	if (opt_piece_mode == MODE_BYTE)
		rc = split_bytes(f);
	else if (opt_piece_mode == MODE_LINE)
		rc = split_lines(f);
	else {
		assert(0);
		rc = 1;
	}

	if (ferror(f)) {
		perror(err_fn);
		rc = 1;
	}

	rc |= close_output();

	if (opt_input_fn != NULL)
		if (fclose(f))
			perror(opt_input_fn);

	return EXIT_FAILURE;
}

int main (int argc, char *argv[])
{
	struct pathelem *pe;
	error_t rc;

	pu_init();

	rc = argp_parse(&argp, argc, argv, 0, NULL, NULL);
	if (rc) {
		fprintf(stderr, _("argp_parse failed: %s\n"), strerror(rc));
		return 1;
	}

	pe = path_split(opt_prefix);
	if ((strlen(pe->basen) + opt_suffix_len) > NAME_MAX) {
		fprintf(stderr, _("prefix + suffix too large\n"));
		return 1;
	}
	path_free(pe);

	return execute_split();
}

